barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt

```diff
@@ -0,0 +1,333 @@
+    private var cameraDetectionDelay = 2000L
+    //The barcode closest to the center of the screen in portrait mode.
+    private var activeBarcode : Barcode? = null
+
+    //Camera preview screen
+    private lateinit var sourcePreview: CameraSourcePreview
+    //The actual camera source
+    private lateinit var source: CameraSource
+    //Overlay drawn over the barcodes
+    private lateinit var overlay: GraphicOverlay<BarcodeGraphic>
+    //Camera preview screen dimensions
+    private lateinit var sourcePreviewMeasurements: Rect
+    private lateinit var snackbarView: View
+
+    //All of the barcodes we read during cameraDetectionDelay.
+    private val readBarcodes = mutableListOf<Barcode>()
+    //Triggered when we read the first barcode. Used to start cameraDetectionDelay countdown.
+    private val hasBarcode = MutableLiveData<Boolean>() //Ensure this variable is only changed ONCE.
+    //To sync barcode detections and avoid dupes.
+    private val lock = Any()
+
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        val view = inflater.inflate(R.layout.fragment_barcode_reader, container, false)
+
+        sourcePreview = view.findViewById(R.id.fragment_barcode_scanner_camera_source_preview)
+        overlay = view.findViewById(R.id.fragment_barcode_scanner_camera_source_preview_overlay)
+        snackbarView = requireActivity().findViewById(android.R.id.content)
+
+        //Gather barcode information.
+        arguments?.getInt(BARCODE_FORMATS_BUNDLE_KEY)?.let { barcodeFormats = it }
+        arguments?.getInt(BARCODE_CAMERA_RESOLUTION_WIDTH_BUNDLE_KEY)?.let { cameraResolutionWidth = it }
+        arguments?.getInt(BARCODE_CAMERA_RESOLUTION_HEIGHT_BUNDLE_KEY)?.let { cameraResolutionHeight = it }
+        arguments?.getInt(BARCODE_STORAGE_MIN_PERCENTAGE_BUNDLE_KEY)?.let { deviceStorageMinPercentage = it }
+        arguments?.getLong(BARCODE_CAMERA_DETECTION_DELAY_BUNDLE_KEY)?.let { cameraDetectionDelay = it }
+        arguments?.getFloat(BARCODE_CAMERA_FRAMES_PER_SECOND_BUNDLE_KEY)?.let { cameraFramesPerSeconds = it }
+
+        //Setup the observables and their handlers.
+        setupObservables()
+
+        //Check permissions before opening the camera
+        val permissionStatus = ActivityCompat.checkSelfPermission(requireContext(), android.Manifest.permission.CAMERA)
+
+        if (permissionStatus == PackageManager.PERMISSION_GRANTED) {
+            //If the permission is granted, we need to wait for the system to measure the camera preview screen.
+            //After we have that information, we need to save it to calculate the closest barcode to the center.
+            sourcePreview.afterMeasured {
+                sourcePreviewMeasurements = Rect(left, top, right, bottom)
+            }
+
+            initCamera()
+        } else {
+            requestCameraPermission()
```

Comment:
Add this as a default behavior but with the possibility of being overriden via a callback.
So if you add a callback for this, it will call it, and the callback will do whatever needed to deal with it, if not, call this.

Reply:
For the whole block (init & request permissions) or just to request permissions?

Reply:
I'd say just the requestPermission, wdyt?

Reply:
Yeah, I agree. Added the callback 👍